exchanges/stream/reporting.go

```diff
@@ -0,0 +1,119 @@
+	return defaultProcessReporterManager{period: time.Minute}
+}
+
+// defaultProcessReporterManager is a default implementation of ProcessReporter
+type defaultProcessReporterManager struct{ period time.Duration }
+
+// New returns a new DefaultProcessReporter instance for a connection
+func (d defaultProcessReporterManager) New(conn Connection) ProcessReporter {
+	reporter := &defaultProcessReporter{ch: make(chan struct{})}
+	go reporter.collectMetrics(conn, d.period)
+	return reporter
+}
+
+// DefaultProcessReporter provides a thread-safe implementation of the ProcessReporter interface.
+// It tracks operation metrics, including the number of operations, average processing time, and peak processing time.
+type defaultProcessReporter struct {
+	operations          int64
+	errors              int64
+	totalProcessingTime time.Duration
+	peakProcessingTime  time.Duration
+	peakCause           []byte
+	ch                  chan struct{}
+	m                   sync.Mutex
+}
+
+// Report logs the processing time for a received data packet and updates metrics.
+func (r *defaultProcessReporter) Report(read time.Time, data []byte, err error) {
+	processingDuration := time.Since(read)
+	r.m.Lock()
+	defer r.m.Unlock()
+	r.operations++
+	if err != nil {
+		r.errors++
+	}
+	r.totalProcessingTime += processingDuration
+	if processingDuration > r.peakProcessingTime {
+		r.peakProcessingTime = processingDuration
+		r.peakCause = data
+	}
+}
+
+// Close closes the process reporter
+func (r *defaultProcessReporter) Close() {
+	r.m.Lock()
+	close(r.ch)
+	r.m.Unlock()
+}
+
+// collectMetrics runs in a separate goroutine to periodically log aggregated metrics.
+func (r *defaultProcessReporter) collectMetrics(conn Connection, period time.Duration) {
+	if period == 0 {
+		panic("period duration for collecting metrics must be greater than 0")
+	}
+	timer := time.NewTimer(time.Until(time.Now().Truncate(period).Add(period)))
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-r.ch:
+			return
+		case <-timer.C:
+			timer.Reset(time.Until(time.Now().Truncate(period).Add(period)))
+			r.m.Lock()
+			if r.operations > 0 {
+				avgOperationsPerSecond := float64(r.operations) / 60
+				avgProcessingTime := r.totalProcessingTime / time.Duration(r.operations)
+				peakTime := r.peakProcessingTime
+				peakCause := r.peakCause
+				errors := r.errors
+				// Reset metrics for the next interval.
+				r.operations, r.totalProcessingTime, r.peakProcessingTime, r.peakCause, r.errors = 0, 0, 0, nil, 0
+				r.m.Unlock()
+				if len(peakCause) > 100 {
+					peakCause = append(peakCause[:100], []byte("...")...)
+				}
+				// Log metrics outside of the critical section to avoid blocking other threads.
+				log.Debugf(log.WebsocketMgr, "Connection: %v Operations/Second: %.2f, Avg Processing/Operation: %v, Errors: %v Peak: %v Cause: %v...", conn.GetURL(), avgOperationsPerSecond, avgProcessingTime, errors, peakTime, string(peakCause))
```

Comment:
```suggestion
					peakCause = append(peakCause[:100], []byte("...")...)
				}
				// Log metrics outside of the critical section to avoid blocking other threads.
				log.Debugf(log.WebsocketMgr, "Connection: %v Operations/Second: %.2f, Avg Processing/Operation: %v, Errors: %v Peak: %v Cause: %v", conn.GetURL(), avgOperationsPerSecond, avgProcessingTime, errors, peakTime, string(peakCause))
```
I removed the ellipses in the main output because you add them if its > 100 and doesn't really mean much if < 100

Reply:
done